### ISSUE TITLE
Add configuration in `java-generator-maven-plugin` to allow setting `generateBuilderPackage` option in sundrio @Buildable annotation

### ISSUE THREAD

**Issue by musabshak** [fabric8io/kubernetes-client#5452]
### Is your enhancement related to a problem? Please describe

The[ `java-generator-maven-plugin`](https://github.com/fabric8io/kubernetes-client/blob/main/doc/java-generation-from-CRD.md) supports an `extraAnnotations` configuration option that generates sundrio @Buildable annotations, but there is no configuration offered for the annotations that are generated. In particular, the `generateBuilderPackage` option in the generated annotations is hard-corded to be false. If @Buildable annotations are used in any other part of the project and `generateBuilderPackage` is set to true there, sundrio throws an exception (ie in entire project, can only use one value for `generateBuilderPackage` -- true or false.

```
 extraAnnotations
      User property: fabric8.java-generator.extra-annotations
      Generate Extra annotation for lombok and sundrio integration
```
```
@Buildable(
    editableEnabled = false,
    validationEnabled = false,
    generateBuilderPackage = false,
    lazyCollectionInitEnabled = false,
    builderPackage = "io.fabric8.kubernetes.api.builder")
```

sundrio docs - https://github.com/sundrio/sundrio/blob/main/annotations/builder/readme.md#eliminating-runtime-dependencies

### Describe the solution you'd like

Add a configuration option in the maven plugin, make code change to respond to that configuration option

**Comment by manusa** [comment]
The template approach doesn't really convince me, especially given the multiple ways in which the generator might be invoked. The resulting code might not look great in some scenarios.

I think the interface approach is more flexible and is open for extension in the future.
We could also use SPI to automatically load implementations provided that they are in the class path. Our code could then be refactored to provide a default implementation (current Sundrio annotation adder)

**Comment by fabiobrz** [comment]
I agree with @andreaTP consideration about not exposing the configuration 1-1, and the rationale for this.
That being said, I also agree with @manusa about the template-based solution being a bit tricky to handle.
The interface based solution seems reasonable to me, the default implementation could do what is done now.
Loading via SPI seems good as well, and a factory implementation for this would default on the base interface if no others are found scanning the classpath.

**Comment by fabiobrz** [comment]
> I agree with @andreaTP consideration about not exposing the configuration 1-1, and the rationale for this. That being said, I also agree with @manusa about the template-based solution being a bit tricky to handle. The interface based solution seems reasonable to me, the default implementation could do what is done now. Loading via SPI seems good as well, and a factory implementation for this would default on the base interface if no others are found when scanning the classpath.

BTW - using `extraAnnotationAdderClass` would make for a dynamic way to load from the classpath, which would not work for the Quarkus use case as well. So it seems to me we're left with 2 options:

1. use `extraAnnotationTemplate` and keep the expected format documented
2. go with `ExtraAnnotationAdder`, use a factory manager to elect a SPI registered factory that will provide a concrete implementation from classpath if any, otherwise default on the one with the default behavior and accept the limitation for this in the Quarkus use case.

**Comment by andreaTP** [comment]
Thanks a lot for the feedback @manusa and @fabiobrz !
I'm trying to form my mind about this and I'm not fully convinced, here are my arguments:

- I completely get the downsides of using a "template string", things can go pretty messy. Opening up this possibility without integrating a templating engine is going to fall short pretty quickly
- I find the SPI approach really great and it will be my go-to choice, but, I think that the user experience is going to be pretty bad. I expect that most of our users are consuming the `java-generator` through the Maven plugin and [adding artifacts to the plugin classpath](https://webtide.com/extending-the-maven-plugin-classpath-at-runtime/) seems to be challenging. The most viable option would be to re-pack the `java-generator-maven-plugin` with the additional class/es, but this solution involves definitely too much boilerplate and busy work.

What do you think?
Am I missing anything?
More alternative options?

**Comment by manusa** [comment]
> [adding artifacts to the plugin classpath](https://webtide.com/extending-the-maven-plugin-classpath-at-runtime/) seems to be challenging. The most viable option would be to re-pack the `java-generator-maven-plugin` with the additional class/es, but this solution involves definitely too much boilerplate and busy work.

What's wrong with the `<plugin><dependencies>` configuration? why won't that work?

We do use this in Eclipse JKube for multiple purposes.

Here is a quickstart leveraging this approach for a custom _Enricher_:

https://github.com/eclipse/jkube/blob/a5fa117e1d91518a1baf8b26c47ccd770a63effc/quickstarts/kit/custom-istio-enricher/app/pom.xml#L80-L87

**Comment by andreaTP** [comment]
Thanks @manusa for the additional resource!

> What's wrong with the <plugin><dependencies> configuration? why won't that work?

Look at the request in this ticket:
- the user wants to change a boolean parameter in an annotation

What he would need to do is to:
- setup a project with `java-generator` as a transitive and provided dependency
- implement the SPI interface
- package and redistribute the additional jar (with all the JIRAs, gpg keys etc. for first-timers)
- use the new implementation by adding it to the classpath

I think that it's simply "too much work" compared to the desired outcome.

**Comment by manusa** [comment]
Well, the user should only need to bootstrap a multimodule project where the first module provides the desired implementation and then depend on it on the second module (then use `install` as opposed to `verify` or `package`).

But agreed, it might be too much of an overhead for such a simple purpose.

**Comment by andreaTP** [comment]
### Another alternative to consider

Add a configuration `extraAnnotationAdderSource` with the following semantics:
- it will be a path to a `.java` source file containing a single class annotated with the desired annotations.

Considerations:
We are already directly using JavaParser, which is great for those kinds of selective syntactic extraction of Java source code.

PRO:
- the user can even have such source file in their `src` folder, this way the annotations will also be compiled by `javac` giving great confidence about the content.

**Comment by manusa** [comment]
> Another alternative to consider

I'm not sure I fully grasp how this work, but does sound like something to consider :+1:

**Comment by andreaTP** [comment]
Instead of asking the user to provide the code to be injected "directly", we can ask to provide a `.java` valid source code file.
If the option is not provided the generation falls back to the hard-coded values.
If the option is specified the generator will:
- parse the `.java` file
- get the first available `class`
- extract the static annotations on the `class` (limitation: they should be "fixed")
- apply the extracted annotations to each and every `class` generated by the `java-generator`

This is some kind of middle ground, as the annotations to add cannot be dependent on the content(it's not a template, just a fixed value); but, being able to define those in a `.java` file opens the space for checks with the compiler etc., making it less error-prone and less fragile.

**Comment by fabiobrz** [comment]
Thanks for reviewing the feedback, @andreaTP - while I understand the explanation you gave about this would work, it seems quite awkward to provide a Java class just in order to let the annotations be extracted, WDYT? Is there any reference for implementations where this practice is used?

Otherwise I'd really vote for something simpler, as a configurable - via property - string.

**Comment by shawkins** [comment]
@fabiobrz @andreaTP specifically to satisfy the Buildable issue the basic workaround is to use a separate module to generate you object model, that way there will only be one value of generateBuilderPackage in play.

It is also possible to log an enhancement against sundrio to allow both true and false values for generateBuilderPackage - as long as all true values match.

**Comment by fabiobrz** [comment]
> @fabiobrz @andreaTP specifically to satisfy the Buildable issue the basic workaround is to use a separate module to generate you object model, that way there will only be one value of generateBuilderPackage in play.
> 
> It is also possible to log an enhancement against sundrio to allow both true and false values for generateBuilderPackage - as long as all true values match.

Thanks @shawkins but still, IIUC this would only solve the conflict issue (i.e. true || false), while we would have to expose a property for configuring such annotation anyway, or am I missing something?

**Comment by shawkins** [comment]
> Thanks @shawkins but still, IIUC this would only solve the conflict issue (i.e. true || false), while we would have to expose a property for configuring such annotation anyway, or am I missing something?

Yes it is only to solve the immediat issue of the conflict.

**Comment by andreaTP** [comment]
We discussed this issue and those are the follow-up actions:

- unblock the user on this specific issue, already done [here](https://github.com/fabric8io/kubernetes-client/issues/5452#issuecomment-1727746875), cc. @musabshak let us know if moving to a separate module doesn't fix the issue
- postpone the development of the more "generic" feature
- meanwhile, demonstrate how those kinds of transformations are achievable with a separate `maven-plugin`

**Comment by musabshak** [comment]
I was able to unblock myself by just editing the generateBuilderPackage value in other parts of the upstream code to `false`. Thought I'd open the bug for completeness' sake - should have marked it as non-urgent. Thanks for looking into it, and suggesting other workarounds (using a separate module etc).

**Comment by andreaTP** [comment]
Thanks for the feedback, much appreciated!
